### PR TITLE
Add retry logic to groovy based api queries (2)

### PR DESCRIPTION
**Pull request by M-Davies**
* Attempt 2 (attempt 1 = #33)

Issue: https://github.com/AdoptOpenJDK/openjdk-build/issues/1873 
Signed-off-by: Morgan Davies <morgan.davies@ibm.com>

### CODE REVIEW

**Comment by M-Davies** [comment]
Draft until tomorrow

**Comment by M-Davies** [comment]
@karianna If you're happy with this, can we merge and I can kick off some test pipelines?

File: src/JobHelper.groovy [hunk 1]
```diff
@@ -62,23 +62,53 @@ class JobHelper {
...
+                // Successful response
+                if (jsonResponse != null) {
+                    context.println "[SUCCESS] We have a response!\n${jsonResponse}"
+                    break
+                } else {
+                    context.println "[RETRYWARNING] API Request was successful but jsonResponse is null. Retrying in 30 seconds..."
+                    context.sleep 30000
+                }
+            } catch (Exception e) {
+                if (count >= attempts) {
```
Comment (karianna) [inline src/JobHelper.groovy:88]: how would count > attempts?

File: src/JobHelper.groovy [hunk 2]
```diff
@@ -62,23 +62,53 @@ class JobHelper {
...
         }
     }
 
     /**
     * Queries the Adopt API for all releases
+    * @param context
     * @return
     */
-    public static getAvailableReleases() { return queryApi("https://api.adoptopenjdk.net/v3/info/available_releases") }
+    public static getAvailableReleases(def context) {
```
Comment (karianna) [inline src/JobHelper.groovy:110]: where is this being called from?

File: src/JobHelper.groovy (hunk 1 above)
Comment (M-Davies) [reply src/JobHelper.groovy:88]: I used >= just as insurance. It should only need to be ==

File: src/JobHelper.groovy (hunk 2 above)
Comment (M-Davies) [reply src/JobHelper.groovy:110]: https://github.com/AdoptOpenJDK/openjdk-build/pull/1893

File: src/JobHelper.groovy [hunk 3]
```diff
@@ -62,23 +62,53 @@ class JobHelper {
...
     */
-    private static queryApi(String query) {
-        try {
-            def get = new URL(query).openConnection()
-            get.setRequestProperty("User-Agent", "adopt-jenkins-helper")
-            def parser = new JsonSlurper()
-            return parser.parseText(get.getInputStream().getText())
-        } catch (Exception e) {
-            throw new RuntimeException("${e.getMessage()}")
+    private static queryApi(String query, Integer attempts, def context) {
```
Comment (johnoliver) [inline src/JobHelper.groovy:69]: The library is pure groovy, as such println should be available without the context, and sleep can be achieved with:
```
try {
    Thread.sleep(30000);
} catch (Exception e) {
    println "Sleep interrupted"
}
```

If you do that you dont need the context

Comment (M-Davies) [reply src/JobHelper.groovy:69]: I thought that too but when I tried it as part of my [first PR](https://github.com/AdoptOpenJDK/openjdk-jenkins-helper/pull/33/files), the println's that I put in didn't spit out any output https://ci.adoptopenjdk.net/job/build-scripts/job/openjdk16-pipeline/10/console

File: src/JobHelper.groovy (hunk 1 above)
Comment (karianna) [reply src/JobHelper.groovy:88]: Right, worth changing that not to confuse the next reader.

Comment (M-Davies) [reply src/JobHelper.groovy]: Done

**Pull request merged**